[JENKINS-47768] - Avoid having "authenticated" twice in the group membership of a user (LastGrantedAuthorities)

Wadeck opened the pull request:
- this occur when the SecurityRealm potentially already grants that role (like in github-oauth-plugin)

See [JENKINS-47768](https://issues.jenkins-ci.org/browse/JENKINS-47768).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Appropriate autotests or explanation to why this change has no tests

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

Wadeck committed:
Avoid having "authenticated" twice in the group membership of a user
- this occur when the SecurityRealm potentially already grants that role (like in github-oauth-plugin)

Wadeck commented:
Instead of checking that on each "update" I think it's better to do on the information retrieval. The main advantage is that we do not bother performance for user that do not use the "cache" that is provided by LastGrantedAuthorities. I think it's the majority of cases.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

oleg-nenashev reviewed (commented):
Would it be possible to use Set or Map instead? Just to avoid N^2 complexity on this method. 

Maybe it's just better to modify the internal data structure and to use Maps in new API

oleg-nenashev commented on core/src/main/java/jenkins/security/LastGrantedAuthoritiesProperty.java:
```diff
@@ -48,14 +48,20 @@ public UserProperty reconfigure(StaplerRequest req, JSONObject form) throws Form
+        }
+
+        String authenticatedRole = SecurityRealm.AUTHENTICATED_AUTHORITY.getAuthority();
+        List<GrantedAuthority> grantedAuthorities = new ArrayList<>();
```
Initialize with `roles.length`

Wadeck replied:
changed

Wadeck committed:
- changed as requested by Oleg
- the list has a maximum of roles.length and in reality it's either roles.length or (roles.length-1), so the maximum is ok

dwnusbaum commented on core/src/main/java/jenkins/security/LastGrantedAuthoritiesProperty.java:
```diff
@@ -48,14 +48,20 @@ public UserProperty reconfigure(StaplerRequest req, JSONObject form) throws Form
+        for (int i = 0; i < roles.length; i++){
+            // to avoid having twice that role
+            if(!authenticatedRole.equals(roles[i])){
+                grantedAuthorities.add(new GrantedAuthorityImpl(roles[i]));
```
If `this.roles` is not null, but doesn't contain `SecurityRealm.AUTHENTICATED_AUTHORITY`, then `SecurityRealm.AUTHENTICATED_AUTHORITY` won't be added to the result. Unless I'm missing something I think you want to add `AUTHENTICATED_AUTHORITY` to the the list before the loop to make sure it's included, right? (And initialize the list to `roles.length + 1`).

oleg-nenashev reviewed (changes requested):
Agreed with @dwnusbaum, the test failures are related

Wadeck committed:
- fix problem of missing the "authenticated" authority

Wadeck committed:
- convert the Groovy script to a Java version
- the Groovy test was not run by default (IIUC Groovy scripts are not compiled if placed in java src folder)

dwnusbaum reviewed (approved):
🐝

oleg-nenashev reviewed (approved):
🐝 and 👍 for reworking just another Groovy test

oleg-nenashev commented:
@reviewbybees done
